# load data (1)

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/msfrv/interviews/main/prices.csv')
df.head()

stock        date  price
0   IEC  2018-01-01  71.36
1   IEC  2018-01-02  72.10
2   IEC  2018-01-03  72.70
3   IEC  2018-01-04  72.54
4   IEC  2018-01-05  72.68

## reformat
Convert the dataframe such that the rows are date, columns are stocks.
The first few rows should be like:

```
              ETU    HLR    IEC    IZP     MDH    NOR    QLY    QWM    YWC    ZCV  
2018-01-01  25.35  54.50  71.36  66.63   97.97  50.64  49.25  64.59  90.21  66.76  
2018-01-02  25.90  53.14  72.10  66.33   96.53  50.80  48.70  68.58  90.47  65.80  
2018-01-03  36.90  54.60  72.70  68.10   98.15  50.55  51.95  68.02  90.16  64.21  
2018-01-04  24.50  54.00  72.54  67.71  103.91  50.48  50.10  66.20  90.47  66.43  
2018-01-05  24.30  54.82  72.68  66.93   95.81  50.41  51.60  62.56  90.33  64.93  
...
```

In [3]:
# code here


## find outliers
find datapoints (stock+date) which are >2 sigmas (standard deviation) away from their means.

Show their z-scores and the prices.

The output should be like:

```
 stock        date    z-score   price
   ETU  2018-01-03   7.596733   36.90
   IEC  2018-03-01  10.474413   91.24
   IEC  2018-07-24 -11.128219   51.56
...
```
note: z-score is $(x-\mu)/\sigma$


In [4]:
# code here


## remove outliers
for each of the outliers found above, replace their prices by interpolating between the values before and after.

For example, for the first outlier (price of ETU at 2018-01-03), the prices before and after that day are:

```
      date   prices
2018-01-02    25.90
2018-01-03    36.90
2018-01-04    24.50
```
We will replace the price of ETU at 2018-01-03 by: $(25.90+24.50)/2=25.20$

Please output the list of overridden prices. The first few should be:
```
stock  date      
ETU    2018-01-03    25.20
IEC    2018-03-01    71.84
       2018-07-24    71.58
MDH    2018-11-09    98.78
...
```

In [5]:
# code here


The prices should now be like this :
```
stock         ETU    HLR    IEC    IZP     MDH    NOR    QLY    QWM    YWC    ZCV
2018-01-01  25.35  54.50  71.36  66.63   97.97  50.64  49.25  64.59  90.21  66.76
2018-01-02  25.90  53.14  72.10  66.33   96.53  50.80  48.70  68.58  90.47  65.80
2018-01-03  25.20  54.60  72.70  68.10   98.15  50.55  51.95  68.02  90.16  64.21
2018-01-04  24.50  54.00  72.54  67.71  103.91  50.48  50.10  66.20  90.47  66.43
2018-01-05  24.30  54.82  72.68  66.93   95.81  50.41  51.60  62.56  90.33  64.93
...
```  

# load weights
This file contains the weights of those 10 stocks in 20 indices

In [6]:
w = pd.read_csv('https://raw.githubusercontent.com/msfrv/interviews/main/weights.csv')
w.head()

IndexName stock    weight
0   Index00   HLR  0.249400
1   Index00   MDH  0.257100
2   Index00   ZCV  0.384000
3   Index01   HLR  1.000000
4   Index02   MDH  0.573095

## compute recomposed returns

Please compute the *recomposed returns* of the indices by:

$ r_{i,d} = \sum_j w_{i,j} (p_{j,d}/p_{j,d-1}-1)$

where 
* $r_{i,d}$ is the recomposed return of the $i^{th}$ Index on date $d$
* $w_{i,j}$ is the weight of the $j^{th}$ stock in the $i^{th}$ Index
* $p_{j,d}$ is the price of the $j^{th}$ stock on date $d$

For example, the recomposed return of Index00 at 2018-01-02 is computed by:

$ 0.2494\times(53.14/54.50-1) + 0.2571\times(96.53/97.97-1) + 0.3840\times(65.80/66.76-1) = -0.01552438$

The result should be a 260x20 (Ndates x Nindices) dataframe. The first few rows should be like:
```
IndexName    Index00   Index01   Index02   Index03   Index04 ...  
2018-01-02 -0.015524 -0.024954 -0.004104 -0.006599  0.007344 ...   
2018-01-03  0.001888  0.027475  0.007906  0.020762  0.007748 ...   
2018-01-04  0.025624 -0.010989  0.033516  0.006294 -0.003932 ...   
...
```

In [7]:
# code here